go/internal/tinder/driver_localdiscovery.go

```diff
@@ -0,0 +1,394 @@
+package tinder
+
+import (
+	"context"
+	"errors"
+	"io"
+	"math"
+	mrand "math/rand"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+	"go.uber.org/zap"
+
+	mcma "berty.tech/berty/v2/go/internal/multipeer-connectivity-transport/multiaddr"
+	ggio "github.com/gogo/protobuf/io"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr-net"
+	msmux "github.com/multiformats/go-multistream"
+)
+
+const recProtocolID = protocol.ID("berty/p2p/localrecord")
+
+type localDiscovery struct {
+	logger       *zap.Logger
+	host         host.Host
+	peerCache    map[string]*pCache
+	peerCacheMux sync.RWMutex
+	rng          *mrand.Rand
+	rngMux       sync.Mutex
+}
+
+type pCache struct {
+	recs map[peer.ID]*pRecord
+	mux  sync.Mutex
+}
+
+type pRecord struct {
+	peer   peer.AddrInfo
+	expire int64
+}
+
+type StreamWrapper struct {
+	network.Stream
+	io.ReadWriter
+}
+
+// LocalDiscovery is a discovery.Discovery
+// https://github.com/libp2p/go-libp2p-core/blob/master/discovery/discovery.go
+var _ discovery.Discovery = (*localDiscovery)(nil)
+
+// LocalDiscovery is a network.Notifiee
+// https://github.com/libp2p/go-libp2p-core/blob/master/network/notifee.go
+var _ network.Notifiee = (*localDiscovery)(nil)
+
+// LocalDiscovery is a Driver
+var _ Driver = (*localDiscovery)(nil)
+
+func NewLocalDiscovery(logger *zap.Logger, host host.Host, rng *mrand.Rand) Driver {
+	ld := &localDiscovery{
+		logger:    logger.Named("tinder/localDiscovery"),
+		host:      host,
+		rng:       rng,
+		peerCache: make(map[string]*pCache),
+	}
+	host.Network().Notify(ld)
+	host.SetStreamHandler(recProtocolID, ld.handleStream)
+	return ld
+}
+
+// Delete expired entries of a CID in the peer cache
+func (ld *localDiscovery) cleanPeerCache(cid string) {
+	currentTime := time.Now().Unix()
+
+	ld.peerCacheMux.RLock()
+	_, ok := ld.peerCache[cid]
+	ld.peerCacheMux.RUnlock()
```

Comment:
if you use this patern a lot you should consider using:
```go
func (ld *localDiscovery) GetCache(key string) (ok bool, c pCache) {
     ld.peerCacheMux.RLock()
     ok, c = ld.peerCache[cid]
     ld.peerCacheMux.RUnlock()
     return
}

Reply:
Done